### ISSUE TITLE
Bug in Unicode surrogate pair conversion in Perl_utf16_to_utf8

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6598]
Migrated from [rt.perl.org#22946](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22946) (status was 'resolved')

Searchable as RT22946$

**Comment by p5pRT** [comment]
### From Tom.Dinger@Scansoft.com

This is a bug report for perl from Tom\.Dinger@&#8203;scansoft\.com\,
generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.


This is a bug against code in the Perl 5\.8\.0 source code for utf8\.c\, that
seems to have been present for some time \-\- probably is a \_very\_ seldomly
used feature&#8203;: Unicode surrogate pairs\.

There is an error in the function Perl\_utf16\_to\_utf8\(\)\, in the file utf8\.c\,
in the logic that converts a surrogate pair in UTF\-16 BE format to the
actual Unicode value\.  The second 16\-bit value is being incompletely
extracted; only a single byte of the two is being used\.

This looks like it is left over from a time when the 'p' parameter to the
function was typed as a 'U16\*'; now it is a 'U8\*'\.  Note that the code
earlier in the function to pick up the first 16\-bit element is correct\.

Here is a very simple unified context\-diff of a fix&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- utf8.c.orig	Mon Jul 01 11:10:18 2002
+++ utf8.c	Tue Jul 01 14:36:12 2003
@@ -828,7 +828,8 @@
 	    continue;
 	}
 	if (uv >= 0xd800 && uv < 0xdbff) {	/* surrogates */
-	    UV low = *p++;
+	    UV low = (p[0] << 8) + p[1];
+	    p += 2;
 	    if (low < 0xdc00 || low >= 0xdfff)
 		Perl_croak(aTHX_ "Malformed UTF-16 surrogate");
 	    uv = ((uv - 0xd800) << 10) + (low - 0xdc00) + 0x10000;

---

```
</details>
Flags:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl v5\.6\.1&#8203;:

Summary of my perl5 \(revision 5 version 6 subversion 1\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=undef usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags ='\-nologo \-O1 \-MD \-DNDEBUG \-DWIN32 \-D\_CONSOLE
\-DNO\_STRICT \-DHAVE\_DES\_FCRYPT  \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS
\-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-O1 \-MD \-DNDEBUG'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=4
&nbsp;    alignbytes=8\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-release
\-libpath&#8203;:"C&#8203;:\\Perl561\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth="C&#8203;:\\Perl561\\lib\\CORE"
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib
uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib
msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib  netapi32\.lib
uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib odbc32\.lib odbccp32\.lib
msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl56\.lib
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-release
\-libpath&#8203;:"C&#8203;:\\Perl561\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;    ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY



---
@&#8203;INC for perl v5\.6\.1&#8203;:
&nbsp;    c&#8203;:\\Perl561\\site\\lib
&nbsp;    c&#8203;:\\Perl561\\lib
&nbsp;    c&#8203;:\\lib\\perl5
&nbsp;    c&#8203;:\\lib\\perl
&nbsp;    \.



---
Environment for perl v5\.6\.1&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
 
PATH=c&#8203;:\\ScanSoft\\ntbin;c&#8203;:\\ScanSoft\\bin;c&#8203;:\\bin\\win32;c&#8203;:\\bin\\nt;c&#8203;:\\bin\\arepa;c
:\\bin\\dos;C&#8203;:\\WINNT\\system32;C&#8203;:\\WINNT;C&#8203;:\\WINNT\\System32\\Wbem
&nbsp;    PERL5LIB=c&#8203;:\\Perl561\\site\\lib;c&#8203;:\\Perl561\\lib;c&#8203;:\\lib\\perl5;c&#8203;:\\lib\\perl
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @rgs

Dinger\, Tom wrote&#8203;:
> 
> This is a bug against code in the Perl 5\.8\.0 source code for utf8\.c\, that
> seems to have been present for some time \-\- probably is a \_very\_ seldomly
> used feature&#8203;: Unicode surrogate pairs\.
> 
> There is an error in the function Perl\_utf16\_to\_utf8\(\)\, in the file utf8\.c\,
> in the logic that converts a surrogate pair in UTF\-16 BE format to the
> actual Unicode value\.  The second 16\-bit value is being incompletely
> extracted; only a single byte of the two is being used\.

Thanks for the report\.
Could you produce a simple test case ? That would be helpful for the
test suite\. \(mail attachments accepted\)

**Comment by p5pRT** [comment]
### From @rgs

Solved by change \#20211\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
